### ISSUE TITLE
SIGSEGV when trying to use Lucene

### ISSUE THREAD

**Issue by jacobobryant** [xtdb/xtdb#1415]
After including `juxt/crux-lucene {:mvn/version "20.12-1.13.0-alpha"}` and calling `start-node` with `:crux.lucene/lucene-store {:db-dir "data/lucene"}`, I get the error below. Here's the hs_err* file in case it's helpful: [hs_err_pid20743.log](https://github.com/juxt/crux/files/5936059/hs_err_pid20743.log)

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f49942bb9c0, pid=20743, tid=20870
#
# JRE version: OpenJDK Runtime Environment (11.0.9.1+1) (build 11.0.9.1+1-Ubuntu-0ubuntu1.20.04)
# Java VM: OpenJDK 64-Bit Server VM (11.0.9.1+1-Ubuntu-0ubuntu1.20.04, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# C  0x00007f49942bb9c0
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/jacob/dev/findka/essays/hs_err_pid20743.log
Compiled method (c1)   22978 9008       3       crux.rocksdb.RocksKv::new_snapshot (58 bytes)
 total in heap  [0x00007f4a29738110,0x00007f4a297399f0] = 6368
 relocation     [0x00007f4a29738288,0x00007f4a297383c8] = 320
 main code      [0x00007f4a297383e0,0x00007f4a297393e0] = 4096
 stub code      [0x00007f4a297393e0,0x00007f4a29739450] = 112
 oops           [0x00007f4a29739450,0x00007f4a29739460] = 16
 metadata       [0x00007f4a29739460,0x00007f4a297394e0] = 128
 scopes data    [0x00007f4a297394e0,0x00007f4a29739768] = 648
 scopes pcs     [0x00007f4a29739768,0x00007f4a297399b8] = 592
 dependencies   [0x00007f4a297399b8,0x00007f4a297399c8] = 16
 nul chk table  [0x00007f4a297399c8,0x00007f4a297399f0] = 40
Compiled method (nm)   22980 9010     n 0       org.rocksdb.RocksDB::getSnapshot (native)
 total in heap  [0x00007f4a30ac1890,0x00007f4a30ac1c38] = 936
 relocation     [0x00007f4a30ac1a08,0x00007f4a30ac1a38] = 48
 main code      [0x00007f4a30ac1a40,0x00007f4a30ac1c30] = 496
 oops           [0x00007f4a30ac1c30,0x00007f4a30ac1c38] = 8
```

**Comment by jacobobryant** [comment]
I see this was discussed on Slack just a few days ago--let me know if a minimal repro example would be helpful.

**Comment by refset** [comment]
Hi @jacobobryant that error is fundamentally caused by RocksDB - was there anything happening around your `start-node` code that might have held on to a reference to a previous node without first having called `(.close node)`? Or were you using `open-db` or `open-q`?

**Comment by jacobobryant** [comment]
No, and no. I made a discovery though: it works if I don't start the http server. after I start the node once without setting `:crux.http-server/server`, on subsequent starts I can include that option and it doesn't crash. (I haven't yet tested to see if it continues to work if I transact some entities).

I tried to do a minimal example, but instead of running into SIGSEGV, I get the error described in https://github.com/juxt/crux/issues/1399. (But only when I also start the http server--which I notice deobald was doing in that issue too).

Perhaps it is the same bug and is fixed by https://github.com/juxt/crux/pull/1406 (is that on the current snapshot version?)--and for some reason, the data in my main project's database just causes a different failure path (?).

In any case it'd probably be good if I tested out that other fix.

**Comment by jarohen** [comment]
Hey @jacobobryant - I'll deploy another snapshot version now that #1406 is in. Could there be a chance that this is related to #1316 - that is, are you seeing this error when the node's closing?

Cheers,

James

[edit: deployed]

**Comment by jacobobryant** [comment]
Great, I'll try that out soon. The error happens during `start-node`, so I'm pretty sure it's unrelated.

**Comment by jacobobryant** [comment]
Just tried this on the minimal example, and it still gives `Lucene store latest tx mismatch`. Here's the code I'm running FYI:

`deps.edn`:
```clojure
{:deps
 {juxt/crux-core        {:mvn/version "dev-SNAPSHOT"}
  juxt/crux-rocksdb     {:mvn/version "dev-SNAPSHOT"}
  juxt/crux-http-server {:mvn/version "dev-SNAPSHOT"}
  juxt/crux-lucene      {:mvn/version "dev-SNAPSHOT"}}}
```

`src/foo.clj`:
```clojure
(ns foo
  (:require
    [clojure.java.io :as io]
    [crux.api :as crux]))

(defn -main []
  (println "starting node")
  (let [node (crux/start-node
               {:crux/index-store {:kv-store {:crux/module 'crux.rocksdb/->kv-store
                                              :db-dir (io/file "data/index")}}
                :rocksdb-golden {:crux/module 'crux.rocksdb/->kv-store
                                 :db-dir (io/file "data/golden")}
                :crux/document-store {:kv-store :rocksdb-golden}
                :crux/tx-log {:kv-store :rocksdb-golden}
                :crux.lucene/lucene-store {:db-dir "data/lucene"}
                :crux.http-server/server {:port 3000}})]
    (println "submitting tx")
    (crux/await-tx node (crux/submit-tx node [[:crux.tx/put {:crux.db/id :foo}]]))
    (println "closing node")
    (.close node)
    (println "exiting")
    (System/exit 0)))
```

Commands:
```
$ clj -M -m foo
starting node
submitting tx
closing node
exiting
$ rm -r data/index/ data/lucene/
$ clj -M -m foo
starting node
Execution error (IllegalStateException) at crux.lucene/validate-lucene-store-up-to-date (lucene.clj:67).
Lucene store latest tx mismatch

Full report at:
/tmp/clojure-3564988320736430494.edn
```

**Comment by refset** [comment]
Thank you for this @jacobobryant I just confirmed the behaviour locally also.

The addition of the HTTP server _should_ be unrelated, but it seems to be kicking up the ~same dependency resolution issue that https://github.com/juxt/crux/pull/1406 attempted to resolve.

It seems there's no SIGSEGV error now though, at least, is that right? I wonder if that could that have been a red herring caused by some error handling when the node failed to start.

**Comment by refset** [comment]
Hmm, I can no longer reproduce this. I ran a `lein sub install` earlier today whilst doing something else then later came back to the terminal where I was looking at this issue and tried to run the steps again. I'm quite certain I did a `lein sub install` when trying to reproduce yesterday also, but perhaps it was still using an old `dev-SNAPSHOT` version and only now has the local m2 cache picked up the change.

@jacobobryant did you try clearing your m2 cache since the new snapshot version was published?

**Comment by jacobobryant** [comment]
> It seems there's no SIGSEGV error now though, at least, is that right? I wonder if that could that have been a red herring caused by some error handling when the node failed to start.

The SIGSEGV error only occurred in my actual project, not the minimal example. It does seem likely to me that it is a red herring. Once I get the minimal example working I'll see if my main project works too.

Noob question--how do you clear the local m2 cache? Is `rm -r ~/.m2/repository/juxt` sufficient for our purposes? I did that, then ran `clj -Stree` and got the following:

```
Downloading: juxt/crux-http-server/dev-SNAPSHOT/maven-metadata.xml from clojars
Downloading: juxt/crux-lucene/dev-SNAPSHOT/maven-metadata.xml from clojars
Downloading: juxt/crux-core/dev-SNAPSHOT/maven-metadata.xml from clojars
Downloading: juxt/crux-http-server/dev-SNAPSHOT/crux-http-server-dev-20210215.182901-5.pom from clojars
Downloading: juxt/crux-rocksdb/dev-SNAPSHOT/maven-metadata.xml from clojars
Downloading: juxt/crux-lucene/dev-SNAPSHOT/crux-lucene-dev-20210215.182952-5.pom from clojars
Downloading: juxt/crux-core/dev-SNAPSHOT/crux-core-dev-20210215.182759-7.pom from clojars
Downloading: juxt/crux-rocksdb/dev-SNAPSHOT/crux-rocksdb-dev-20210215.182829-5.pom from clojars
Downloading: net/minidev/json-smart/maven-metadata.xml from central
Downloading: juxt/crux-lucene/dev-SNAPSHOT/crux-lucene-dev-20210215.182952-5.jar from clojars
Downloading: juxt/crux-http-server/dev-SNAPSHOT/crux-http-server-dev-20210215.182901-5.jar from clojars
Downloading: juxt/crux-core/dev-SNAPSHOT/crux-core-dev-20210215.182759-7.jar from clojars
Downloading: juxt/crux-rocksdb/dev-SNAPSHOT/crux-rocksdb-dev-20210215.182829-5.jar from clojars
...
```

But I still get `Lucene store latest tx mismatch`.

**Comment by refset** [comment]
Yep `rm -r ~/.m2/repository/juxt` is exactly what I meant :)

I'll try again to reproduce the issue using your snippet on another machine or VM. Just in case it's relevant though, please can you confirm your Clojure version and also that your OS+JRE details from the initial comment are still accurate:
> JRE version: OpenJDK Runtime Environment (11.0.9.1+1) (build 11.0.9.1+1-Ubuntu-0ubuntu1.20.04)

**Comment by jacobobryant** [comment]
Sounds good. And yep, OS and JRE are the same. I'm using Ubuntu via WSL2 on Windows 10 FYI.

```
$ clj --help
Version: 1.10.2.774

$ java -version
openjdk version "11.0.9.1" 2020-11-04
OpenJDK Runtime Environment (build 11.0.9.1+1-Ubuntu-0ubuntu1.20.04)
OpenJDK 64-Bit Server VM (build 11.0.9.1+1-Ubuntu-0ubuntu1.20.04, mixed mode, sharing)
```

**Comment by jarohen** [comment]
> The error happens during `start-node`, so I'm pretty sure it's unrelated.

Ah, sorry, didn't occur to me originally - if there's a Lucene error during startup, `start-node` will then try to shut everything down, which could then trigger whatever's causing the 'segfault on node shutdown' issue.

If that's the case, current theory is that this is then unrelated to the Lucene issue - we can close this one as a duplicate of #1316, and ensure the Lucene issue is tracked separately.

**Comment by refset** [comment]
I just booted up a fresh Ubuntu 20.04 image in VirtualBox with the Clojure 1.10.2.796 and OpenJDK-11 and the various Crux 1.15.0 deps - it all worked fine for me...so I'm not sure how else to try reproducing this. (I don't have access to a Windows 10 machine)

One other cache you might need to `rm -rf` is the `.cpcache` directory inside your project directory - did you already try killing that as well as the `~/.m2/repository/juxt` directories?

**Comment by jarohen** [comment]
> The error happens during `start-node`, so I'm pretty sure it's unrelated.

Ah, sorry, didn't occur to me originally - if there's a Lucene error during startup, `start-node` will then try to shut everything down, which could then trigger whatever's causing the 'segfault on node shutdown' issue.

If that's the case, current theory is then that the segfault is then unrelated to the Lucene issue - we can close this one as a duplicate of #1316, and ensure the Lucene issue is tracked separately.

**Comment by jacobobryant** [comment]
I just tried removing `.cpcache` as well, to no avail. I was able to reproduce the error on a fresh DigitalOcean droplet (Ubuntu 20.04) by cloning [this repo](https://github.com/jacobobryant/crux-lucene-repro) followed by `cd crux-lucene-repro; ./run.sh`. Could you see if that reproduces for you? (I added some setup commands to `run.sh` but didn't test it on another new droplet; but I'm 98% sure I didn't botch anything--adding this warning to account for the 2%)

> Ah, sorry, didn't occur to me originally - if there's a Lucene error during startup, start-node will then try to shut everything down, which could then trigger whatever's causing the 'segfault on node shutdown' issue.
>
> If that's the case, current theory is then that the segfault is then unrelated to the Lucene issue - we can close this one as a duplicate of #1316, and ensure the Lucene issue is tracked separately.

Sounds good. Perhaps re-open #1399 and move this discussion there?

**Comment by refset** [comment]
@jacobobryant Thanks for the highly-convenient repo, I just used it to trivially reproduce the issue on DO myself.

The script pulled down `crux-core-dev-20210215.182759-7.jar` & friends, which in theory should have included the fix for https://github.com/juxt/crux/issues/1399 but they didn't for some reason - it might have been the case that we released these snapshots from a branch that wasn't rebased on `master` at the time, but we can't now say for certain what happened. Opening up `crux-lucene-dev-20210215.182952-5.jar` and looking at the source directly confirms that `lucene.clj` _doesn't_ contain the fix in the `(defn ->lucene-store ...` (i.e. `::sys/before #{[:crux/tx-ingester]}`).

In any case, if I modify the `deps.edn` from your repo to use the newer `1.15` versions then everything seems fine on DO also (note that we've not yet created any newer `dev-SNAPSHOT` releases, so `1.15` is the most recent release there is). Let us know on #1399 if you still get the error, but for now I think we are good to close this issue.

We will have a think to see if there is some way to bake the git sha into the `dev-SNAPSHOT` releases to avoid such situations in future...sorry for the confusion and thanks again for helping us get to the bottom of it!
